<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Analysis • TGVIS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Analysis">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.0/css/all.min.css">
<style>
  body { padding-top: 4.5rem; }
  @media (max-width: 768px) { body { padding-top: 5.5rem; } }
  .col-md-8 { max-width: 90%; }
  .container { max-width: 1400px; }
  .page-main { font-size: 1.05rem; line-height: 1.7; }
</style>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TGVIS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-user-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">User Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-user-guide">
<li><a class="dropdown-item" href="../articles/Introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/Data-Preparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/Analysis.html">Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Statistical-Principle.html">Statistical Principle</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/harryyiheyang/TGVIS"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Analysis</h1>
            
      

      <div class="d-none name"><code>Analysis.Rmd</code></div>
    </div>

    
    
<p>We show an example with step-by-step codes.</p>
<div class="section level2">
<h2 id="load-outcome-gwas-data">Load outcome GWAS data<a class="anchor" aria-label="anchor" href="#load-outcome-gwas-data"></a>
</h2>
<p>Our first step is to load the GWAS summary data after SBayesRC
QC:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://r-datatable.com" class="external-link">data.table</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/apache/arrow/" class="external-link">arrow</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">ldscR</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://harryyiheyang.github.io/TGVIS/">TGVIS</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://glue.tidyverse.org/" class="external-link">glue</a></span><span class="op">)</span></span>
<span><span class="va">LDL</span><span class="op">=</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{filepath}/LDL.ma"</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setDT.html" class="external-link">setDT</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span></span>
<span><span class="va">LDL</span></span>
<span><span class="co">#&gt;                  SNP     A1     A2     freq             b          se         p</span></span>
<span><span class="co">#&gt;               &lt;char&gt; &lt;char&gt; &lt;char&gt;    &lt;num&gt;         &lt;num&gt;       &lt;num&gt;     &lt;num&gt;</span></span>
<span><span class="co">#&gt;       1:  rs12132974      T      C 0.079275 -0.0004260400 0.002584074 0.8690452</span></span>
<span><span class="co">#&gt;       2:  rs12134490      C      A 0.079225 -0.0004651126 0.002584819 0.8571996</span></span>
<span><span class="co">#&gt;       3:  rs17276806      T      C 0.079300 -0.0003976603 0.002583702 0.8776799</span></span>
<span><span class="co">#&gt;       4: rs139867617      T      C 0.079275 -0.0003797266 0.002584074 0.8831724</span></span>
<span><span class="co">#&gt;       5:   rs7526310      T      C 0.123250  0.0011156130 0.002123766 0.5993751</span></span>
<span><span class="co">#&gt;      ---                                                                       </span></span>
<span><span class="co">#&gt; 7356514:  rs28729663      A      G 0.136000 -0.0007953800 0.002315470 0.7312171</span></span>
<span><span class="co">#&gt; 7356515:   rs9616832      C      T 0.073875 -0.0001726575 0.002668717 0.9484154</span></span>
<span><span class="co">#&gt; 7356516: rs147475742      A      G 0.041625  0.0022210817 0.003494952 0.5250949</span></span>
<span><span class="co">#&gt; 7356517: rs115055839      C      T 0.073975 -0.0001415992 0.002667057 0.9576587</span></span>
<span><span class="co">#&gt; 7356518:   rs9616985      C      T 0.073850 -0.0002142612 0.002669133 0.9360196</span></span>
<span><span class="co">#&gt;                N    r2</span></span>
<span><span class="co">#&gt;            &lt;int&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt;       1: 1228642     1</span></span>
<span><span class="co">#&gt;       2: 1228642     1</span></span>
<span><span class="co">#&gt;       3: 1228642     1</span></span>
<span><span class="co">#&gt;       4: 1228642     1</span></span>
<span><span class="co">#&gt;       5: 1228642     1</span></span>
<span><span class="co">#&gt;      ---              </span></span>
<span><span class="co">#&gt; 7356514: 1203805     1</span></span>
<span><span class="co">#&gt; 7356515: 1228648     0</span></span>
<span><span class="co">#&gt; 7356516: 1228648     0</span></span>
<span><span class="co">#&gt; 7356517: 1228648     0</span></span>
<span><span class="co">#&gt; 7356518: 1228648     0</span></span></code></pre></div>
<p>This GWAS summary dataset was derived from the EUR-specific LDL-C
GWAS of the GLGC consortium and subsequently imputed using SBayesRC. The
imputed version is available <a href="https://www.dropbox.com/scl/fi/h49jbm4uoc9fqlrmx70is/LDL.ma?rlkey=zmwp6cwait5c22rpbc18878vj&amp;st=jqzz20u1&amp;dl=0" class="external-link">here</a>.
Every GWAS summary data that has passed SBayesRC QC has exactly
7,356,518 SNPs, and the A1 and A2 alleles of each SNP are consistent
with the A1 and A2 alleles in the SBayesRC LD matrix. The SNP
information file is:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">V</span><span class="op">=</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/fread.html" class="external-link">fread</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{filepath}/ukbEUR_Imputed/snp.info"</span><span class="op">)</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span>SNP<span class="op">=</span><span class="va">ID</span>,<span class="va">A1</span>,<span class="va">A2</span>,CHR<span class="op">=</span><span class="va">Chrom</span>,BP<span class="op">=</span><span class="va">PhysPos</span>,Freq<span class="op">=</span><span class="va">A1Freq</span>,<span class="va">Block</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setDT.html" class="external-link">setDT</a></span><span class="op">(</span><span class="va">.</span><span class="op">)</span></span>
<span><span class="va">V</span></span>
<span><span class="co">#&gt;                  SNP     A1     A2   CHR       BP     Freq Block</span></span>
<span><span class="co">#&gt;               &lt;char&gt; &lt;char&gt; &lt;char&gt; &lt;int&gt;    &lt;int&gt;    &lt;num&gt; &lt;int&gt;</span></span>
<span><span class="co">#&gt;       1:  rs12132974      T      C     1   801661 0.079275     1</span></span>
<span><span class="co">#&gt;       2:  rs12134490      C      A     1   801680 0.079225     1</span></span>
<span><span class="co">#&gt;       3:  rs17276806      T      C     1   801858 0.079300     1</span></span>
<span><span class="co">#&gt;       4: rs139867617      T      C     1   802856 0.079275     1</span></span>
<span><span class="co">#&gt;       5:   rs7526310      T      C     1   804759 0.123250     1</span></span>
<span><span class="co">#&gt;      ---                                                        </span></span>
<span><span class="co">#&gt; 7356514:  rs28729663      A      G    22 51219006 0.137750   591</span></span>
<span><span class="co">#&gt; 7356515:   rs9616832      C      T    22 51219387 0.073875   591</span></span>
<span><span class="co">#&gt; 7356516: rs147475742      A      G    22 51219704 0.041625   591</span></span>
<span><span class="co">#&gt; 7356517: rs115055839      C      T    22 51221731 0.073975   591</span></span>
<span><span class="co">#&gt; 7356518:   rs9616985      C      T    22 51229805 0.073850   591</span></span></code></pre></div>
<p>Here, the column <code>Block</code> is necessary as the LD matrices
are stored block-by-block.</p>
<p>The next step is 1) adding Z-score into this file and 2) zoom in the
target locus (PCSK9) region:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">LDL</span><span class="op">$</span><span class="va">CHR</span><span class="op">=</span><span class="va">V</span><span class="op">$</span><span class="va">CHR</span></span>
<span><span class="va">LDL</span><span class="op">$</span><span class="va">BP</span><span class="op">=</span><span class="va">V</span><span class="op">$</span><span class="va">BP</span></span>
<span><span class="va">LDL</span><span class="op">$</span><span class="va">Block</span><span class="op">=</span><span class="va">V</span><span class="op">$</span><span class="va">Block</span></span>
<span><span class="va">LDL</span><span class="op">$</span><span class="va">Zscore</span><span class="op">=</span><span class="va">LDL</span><span class="op">$</span><span class="va">b</span><span class="op">/</span><span class="va">LDL</span><span class="op">$</span><span class="va">se</span></span>
<span><span class="va">LDL_Locus</span><span class="op">=</span><span class="va">LDL</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">CHR</span><span class="op">==</span><span class="fl">1</span><span class="op">&amp;</span><span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">BP</span><span class="op">-</span><span class="fl">55.5e6</span><span class="op">)</span><span class="op">&lt;</span><span class="fl">0.5e6</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">BP</span><span class="op">)</span></span></code></pre></div>
<p>Here for demonstration purposes, we consider a region with a radius
of 0.5MB. However, in practice, users can flexibly choose the radius
according to specific situations. We do not recommend an excessively
large radius (e.g., 1.5MB) because at that point the dimension of the LD
matrix will typically exceed 10 thousand, which will cause memory
issues.</p>
<p>The next step is estimating the per-SNP heritability and inflation
factor using LDSC:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fitldsc</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/pkg/ldscR/man/ldsc.univ.html" class="external-link">ldsc.univ</a></span><span class="op">(</span>gwas<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">LDL</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>,<span class="st">"Zscore"</span>,<span class="st">"N"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span>,LDSC<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">Hapmap3_EURLDSC</span><span class="op">)</span>,sampling.time<span class="op">=</span><span class="fl">0</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Processing data"</span></span>
<span><span class="co">#&gt; Time difference of 4.571996 secs</span></span>
<span><span class="co">#&gt; [1] "Initial Heritability Estimate"</span></span>
<span><span class="co">#&gt; Time difference of 0.03129983 secs</span></span>
<span><span class="co">#&gt; [1] "Heritability Estimation (Reweighted)"</span></span>
<span><span class="co">#&gt; Time difference of 0.9223781 secs</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">fitldsc</span><span class="op">)</span></span>
<span><span class="co">#&gt;   intercept intercept.se         h2 h2.se       M</span></span>
<span><span class="co">#&gt; 1  1.784062           NA 0.07295848    NA 1094541</span></span>
<span><span class="va">var_inf</span><span class="op">=</span><span class="va">fitldsc</span><span class="op">$</span><span class="va">h2</span><span class="op">/</span><span class="va">fitldsc</span><span class="op">$</span><span class="va">M</span></span></code></pre></div>
<p>We found that our functions sometimes fail due to subtle differences
between data.table and data.frame. Therefore, when encountering issues,
you can convert data.table to data.frame, just as we did when performing
LDSC. We used different LDSC panels to estimate the inflation factor and
heritability for GLGC lipids, and obtained similar results: large
inflation (exceeding 1.5) while heritability is less than 0.1. This is
slightly abnormal. However, we still recommend using LDSC estimates when
there is no evidence to prove this is incorrect.</p>
</div>
<div class="section level2">
<h2 id="load-xqtl-direct-effect-data">Load xQTL direct effect data<a class="anchor" aria-label="anchor" href="#load-xqtl-direct-effect-data"></a>
</h2>
<p>We utilize the fine-mapping results of eQTL data based on v10 data
published by GTEx. Although GTEx also published fine-mapping results for
sQTL and apaQTL, these results only contain credible sets of causal
xQTLs but lack effect sizes, making them temporarily unavailable. We
will update if they provide sQTL and apaQTL with effect sizes. We first
examine the data structure:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path</span><span class="op">=</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{filepath}/GTEx_v10_SuSiE_eQTL"</span><span class="op">)</span></span>
<span><span class="va">parquet_files</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span>path<span class="op">=</span><span class="va">path</span>,pattern<span class="op">=</span><span class="st">"\\.parquet$"</span>,full.names<span class="op">=</span><span class="cn">F</span><span class="op">)</span></span>
<span><span class="va">Tissue</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/strsplit.html" class="external-link">strsplit</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/basename.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">parquet_files</span><span class="op">)</span>,<span class="st">"\\."</span><span class="op">)</span>,<span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op"><a href="https://magrittr.tidyverse.org/reference/pipe.html" class="external-link">%&gt;%</a></span><span class="fu"><a href="https://rdrr.io/r/base/sort.html" class="external-link">sort</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{path}/{Tissue[1]}*.parquet"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="co">#&gt;  [1] "phenotype_id" "gene_name"    "biotype"      "variant_id"   "pip"         </span></span>
<span><span class="co">#&gt;  [6] "af"           "cs_id"        "cs_size"      "afc"          "afc_se"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 6 × 10</span></span></span>
<span><span class="co">#&gt;   phenotype_id   gene_name biotype variant_id    pip     af cs_id cs_size    afc</span></span>
<span><span class="co">#&gt;   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>          <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>     <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>   <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>  <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">1</span> ENSG000002272… WASH7P    unproc… chr1_6476… 0.993  0.067<span style="text-decoration: underline;">5</span>     1       1  0.578</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">2</span> ENSG000002689… ENSG0000… proces… chr1_1010… 0.952  0.012<span style="text-decoration: underline;">0</span>     1       1  0.857</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">3</span> ENSG000002794… WASH9P    unproc… chr1_5991… 0.999  0.041<span style="text-decoration: underline;">5</span>     1       1 -<span style="color: #BB0000;">0.615</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">4</span> ENSG000002284… ENSG0000… transc… chr1_2851… 1      0.016<span style="text-decoration: underline;">9</span>     1       1 -<span style="color: #BB0000;">7.97</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">5</span> ENSG000002284… ENSG0000… transc… chr1_2867… 1.000  0.165      2       1  1.06 </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">6</span> ENSG000002293… CICP3     proces… chr1_7300… 0.034<span style="text-decoration: underline;">4</span> 0.030<span style="text-decoration: underline;">4</span>     1       6 <span style="color: #BB0000;">NA</span>    </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 1 more variable: afc_se &lt;dbl&gt;</span></span></span></code></pre></div>
<p>We found that the first column is <code>Ensemble</code> ID, the
second column is <code>Symbol</code>, and the fourth column is crucial:
it provides <code>CHR_BP_A2_A1_Build</code>, which we will use to match
with the GWAS summary data. Furthermore, each gene may have one or
multiple credible sets, indicated in the <code>cs</code> column. For
eQTLs within a credible set of a gene, GTEx only records the
<code>afc</code> and <code>afc_se</code> corresponding to the eQTL with
the maximum PIP. Since GTEx’s genotype data largely overlaps with
SBayesRC’s 7M SNP data but some portions are still different, we
performed the following QC process. First, SBayesRC’s build is BP37,
while GTEx is BP38, so our lab created a correspondence file ourselves,
which can be downloaded from <a href="https://www.dropbox.com/scl/fi/jhnxxxtj44iyg2mpx7911/7M_BP_3738.parquet?rlkey=lt8hqmb3dxq1o695aam14g51z&amp;st=tns272h5&amp;dl=0" class="external-link">Dropbox</a>:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">V38</span><span class="op">=</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{filepath}/7M_BP_3738.parquet"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">V38</span><span class="op">=</span><span class="va">V38</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">V38</span><span class="op">$</span><span class="va">CHR</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">CHR</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co"># Reduce the size of V38 to improve computational efficiency  </span></span>
<span><span class="va">V38</span><span class="op">$</span><span class="va">MarkerName</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="va">V38</span><span class="op">$</span><span class="va">MarkerName</span><span class="op">)</span></span></code></pre></div>
<p>Then we harmonize this eQTL data:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">a</span><span class="op">$</span><span class="va">phenotype_id</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"\\..*"</span>,<span class="st">""</span>,<span class="va">a</span><span class="op">$</span><span class="va">phenotype_id</span><span class="op">)</span> <span class="co"># remove version subscript</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setDT.html" class="external-link">setDT</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR"</span>,<span class="st">"BP38"</span>,<span class="st">"A2"</span>,<span class="st">"A1"</span>,<span class="st">"Build"</span><span class="op">)</span><span class="op">:=</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/tstrsplit.html" class="external-link">tstrsplit</a></span><span class="op">(</span><span class="va">variant_id</span>,<span class="st">"_"</span>,fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># Extract CHR and other information from the column variant_id</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="va">a</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">CHR</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">CHR</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="co"># Reduce the size of eQTL data to improve computational efficiency  </span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">MarkerName</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">CHR</span>,<span class="st">":"</span>,<span class="va">a</span><span class="op">$</span><span class="va">BP38</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">CHR</span><span class="op">=</span><span class="cn">NULL</span></span></code></pre></div>
<p>Our next step is our additional quality control. We first impute the
afc and afc_se that are set as NA in this data with the afc and afc_se
of the xQTL with the maximum PIP in the same credible set. This
imputation is reasonable because variants within a credible set
theoretically share the same single effect. After that, we merge this
eQTL data with V38. This process may lose some key SNPs, so we performed
additional QC. For the merged data, we retain the eQTL with the maximum
PIP within each credible set.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Impute NA values with the first non-NA afc and afc_se within each credible set</span></span>
<span><span class="va">a</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"afc"</span>,<span class="st">"afc_se"</span><span class="op">)</span><span class="op">:=</span><span class="fu">.</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">afc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">afc_se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,by<span class="op">=</span><span class="fu">.</span><span class="op">(</span><span class="va">phenotype_id</span>,<span class="va">cs_id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="co"># Merge eQTL data with V38 reference data</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">V38</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>,<span class="st">"MarkerName"</span>,<span class="st">"CHR"</span>,<span class="st">"BP"</span><span class="op">)</span><span class="op">]</span>,by<span class="op">=</span><span class="st">"MarkerName"</span><span class="op">)</span></span>
<span><span class="co"># Keep only the eQTL with maximum PIP for each credible set</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="va">a</span><span class="op">[</span><span class="va">a</span><span class="op">[</span>,<span class="va">.I</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pip</span><span class="op">)</span><span class="op">]</span>,by<span class="op">=</span><span class="fu">.</span><span class="op">(</span><span class="va">phenotype_id</span>,<span class="va">cs_id</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">V1</span><span class="op">]</span></span></code></pre></div>
<p>The subsequent steps are standard processing procedures:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Calculate Z-score</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">Zscore</span><span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">afc</span><span class="op">/</span><span class="va">a</span><span class="op">$</span><span class="va">afc_se</span></span>
<span><span class="co"># Remove missing values</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="co"># Harmonize alleles with reference panel</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu">TGVIS</span><span class="fu">::</span><span class="fu"><a href="../reference/allele_harmonise.html">allele_harmonise</a></span><span class="op">(</span>ref_panel<span class="op">=</span><span class="va">LDL_Locus</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>,<span class="st">"A1"</span>,<span class="st">"A2"</span><span class="op">)</span><span class="op">]</span>,<span class="va">a</span><span class="op">)</span></span>
<span><span class="co"># Sort by gene name, base pair position, and absolute Z-score</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">gene_name</span>,<span class="va">BP</span>,<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Zscore</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co"># Select and rename columns</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">SNP</span>,<span class="va">A1</span>,<span class="va">A2</span>,<span class="va">Zscore</span>,Gene<span class="op">=</span><span class="va">phenotype_id</span>,GeneSymbol<span class="op">=</span><span class="va">gene_name</span><span class="op">)</span></span>
<span><span class="co"># Add tissue information</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">Tissue</span><span class="op">=</span><span class="va">Tissue</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># Create a combined variable identifier</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">Variable</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">GeneSymbol</span>,<span class="st">"*"</span>,<span class="va">a</span><span class="op">$</span><span class="va">Tissue</span><span class="op">)</span></span>
<span><span class="va">a</span></span>
<span><span class="co">#&gt; Key: &lt;SNP&gt;</span></span>
<span><span class="co">#&gt;           SNP     A1     A2    Zscore            Gene GeneSymbol</span></span>
<span><span class="co">#&gt;        &lt;char&gt; &lt;char&gt; &lt;char&gt;     &lt;num&gt;          &lt;char&gt;     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1:   rs398038      T      A -2.167417 ENSG00000184313      MROH7</span></span>
<span><span class="co">#&gt; 2:   rs684473      G      C  5.398289 ENSG00000184313      MROH7</span></span>
<span><span class="co">#&gt; 3: rs45542642      A      G -2.408018 ENSG00000184313      MROH7</span></span>
<span><span class="co">#&gt; 4: rs34232196      T      C -4.084504 ENSG00000169174      PCSK9</span></span>
<span><span class="co">#&gt; 5: rs72907991      G      C  9.721504 ENSG00000243725       TTC4</span></span>
<span><span class="co">#&gt; 6:   rs683880      C      T  6.201287 ENSG00000162402      USP24</span></span>
<span><span class="co">#&gt;                  Tissue                   Variable</span></span>
<span><span class="co">#&gt;                  &lt;char&gt;                     &lt;char&gt;</span></span>
<span><span class="co">#&gt; 1: Adipose_Subcutaneous MROH7*Adipose_Subcutaneous</span></span>
<span><span class="co">#&gt; 2: Adipose_Subcutaneous MROH7*Adipose_Subcutaneous</span></span>
<span><span class="co">#&gt; 3: Adipose_Subcutaneous MROH7*Adipose_Subcutaneous</span></span>
<span><span class="co">#&gt; 4: Adipose_Subcutaneous PCSK9*Adipose_Subcutaneous</span></span>
<span><span class="co">#&gt; 5: Adipose_Subcutaneous  TTC4*Adipose_Subcutaneous</span></span>
<span><span class="co">#&gt; 6: Adipose_Subcutaneous USP24*Adipose_Subcutaneous</span></span></code></pre></div>
<p>We write the above QC process as a for loop to perform similar
processing for all 50 GTEx tissues:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">eQTL</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span><span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">Tissue</span><span class="op">)</span><span class="op">)</span><span class="op">{</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://arrow.apache.org/docs/r/reference/read_parquet.html" class="external-link">read_parquet</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/Sys.glob.html" class="external-link">Sys.glob</a></span><span class="op">(</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{path}/{Tissue[i]}*.parquet"</span><span class="op">)</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">phenotype_id</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">sub</a></span><span class="op">(</span><span class="st">"\\..*"</span>,<span class="st">""</span>,<span class="va">a</span><span class="op">$</span><span class="va">phenotype_id</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setDT.html" class="external-link">setDT</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CHR"</span>,<span class="st">"BP38"</span>,<span class="st">"A2"</span>,<span class="st">"A1"</span>,<span class="st">"Build"</span><span class="op">)</span><span class="op">:=</span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/tstrsplit.html" class="external-link">tstrsplit</a></span><span class="op">(</span><span class="va">variant_id</span>,<span class="st">"_"</span>,fixed<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="va">a</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.html" class="external-link">which</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">CHR</span><span class="op">==</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"chr"</span>,<span class="fu"><a href="https://rdrr.io/r/base/unique.html" class="external-link">unique</a></span><span class="op">(</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">CHR</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,<span class="op">]</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">MarkerName</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">CHR</span>,<span class="st">":"</span>,<span class="va">a</span><span class="op">$</span><span class="va">BP38</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">CHR</span><span class="op">=</span><span class="cn">NULL</span></span>
<span><span class="fu"><a href="https://rdatatable.gitlab.io/data.table/reference/setDT.html" class="external-link">setDT</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"afc"</span>,<span class="st">"afc_se"</span><span class="op">)</span><span class="op">:=</span><span class="fu">.</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">afc</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">afc_se</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span>,by<span class="op">=</span><span class="fu">.</span><span class="op">(</span><span class="va">phenotype_id</span>,<span class="va">cs_id</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">V38</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>,<span class="st">"MarkerName"</span>,<span class="st">"CHR"</span>,<span class="st">"BP"</span><span class="op">)</span><span class="op">]</span>,by<span class="op">=</span><span class="st">"MarkerName"</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="va">a</span><span class="op">[</span><span class="va">a</span><span class="op">[</span>,<span class="va">.I</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/which.min.html" class="external-link">which.max</a></span><span class="op">(</span><span class="va">pip</span><span class="op">)</span><span class="op">]</span>,by<span class="op">=</span><span class="fu">.</span><span class="op">(</span><span class="va">phenotype_id</span>,<span class="va">cs_id</span><span class="op">)</span><span class="op">]</span><span class="op">$</span><span class="va">V1</span><span class="op">]</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">Zscore</span><span class="op">=</span><span class="va">a</span><span class="op">$</span><span class="va">afc</span><span class="op">/</span><span class="va">a</span><span class="op">$</span><span class="va">afc_se</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu">TGVIS</span><span class="fu">::</span><span class="fu"><a href="../reference/allele_harmonise.html">allele_harmonise</a></span><span class="op">(</span>ref_panel<span class="op">=</span><span class="va">LDL_Locus</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>,<span class="st">"A1"</span>,<span class="st">"A2"</span><span class="op">)</span><span class="op">]</span>,<span class="va">a</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">gene_name</span>,<span class="va">BP</span>,<span class="fu"><a href="https://rdrr.io/r/base/MathFun.html" class="external-link">abs</a></span><span class="op">(</span><span class="va">Zscore</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">=</span><span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">a</span>,<span class="va">SNP</span>,<span class="va">A1</span>,<span class="va">A2</span>,<span class="va">Zscore</span>,Gene<span class="op">=</span><span class="va">phenotype_id</span>,GeneSymbol<span class="op">=</span><span class="va">gene_name</span><span class="op">)</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">Tissue</span><span class="op">=</span><span class="va">Tissue</span><span class="op">[</span><span class="va">i</span><span class="op">]</span></span>
<span><span class="va">a</span><span class="op">$</span><span class="va">Variable</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="va">a</span><span class="op">$</span><span class="va">GeneSymbol</span>,<span class="st">"*"</span>,<span class="va">a</span><span class="op">$</span><span class="va">Tissue</span><span class="op">)</span></span>
<span><span class="va">eQTL</span><span class="op">[[</span><span class="va">i</span><span class="op">]</span><span class="op">]</span><span class="op">=</span><span class="va">a</span></span>
<span><span class="op">}</span></span>
<span><span class="va">eQTL</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/do.call.html" class="external-link">do.call</a></span><span class="op">(</span><span class="va">rbind</span>,<span class="va">eQTL</span><span class="op">)</span></span>
<span><span class="va">eQTL</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/merge.html" class="external-link">merge</a></span><span class="op">(</span><span class="va">eQTL</span>,<span class="va">LDL_Locus</span><span class="op">[</span>,<span class="st">"SNP"</span><span class="op">]</span>,by<span class="op">=</span><span class="st">"SNP"</span>,all.y<span class="op">=</span><span class="cn">T</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="load-ld-matrix">Load LD matrix<a class="anchor" aria-label="anchor" href="#load-ld-matrix"></a>
</h2>
<p>We utilized the <code>readEig</code> function from SBayesRC to write
a function called <code>build_LD_matrix</code> for loading the LD
matrix. This function can use the block information in
<code>GWAS_Locus</code> to locate the blocks that need to be read, and
then extract the LD matrix for the corresponding SNPs:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">R</span><span class="op">=</span><span class="fu"><a href="../reference/build_LD_matrix.html">build_LD_matrix</a></span><span class="op">(</span>GWAS_Locus<span class="op">=</span><span class="va">LDL_Locus</span>,ldDir<span class="op">=</span><span class="fu"><a href="https://glue.tidyverse.org/reference/glue.html" class="external-link">glue</a></span><span class="op">(</span><span class="st">"{filepath}/ukbEUR_Imputed/"</span><span class="op">)</span>,snpinfo<span class="op">=</span><span class="va">V</span><span class="op">)</span></span>
<span><span class="va">R</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/stats/cor.html" class="external-link">cov2cor</a></span><span class="op">(</span><span class="va">R</span><span class="op">)</span></span></code></pre></div>
<p>We also provide the codes to calculate the LD matrix using all
European individuals in the UKBB:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" tabindex="-1"></a><span class="fu">library</span>(BEDMatrix)</span>
<span id="cb12-2"><a href="#cb12-2" tabindex="-1"></a>EURIND<span class="ot">=</span><span class="fu">readRDS</span>(<span class="fu">glue</span>(<span class="st">"{filepath}/BED_file/UKBB_EURIND.rds)"</span>)</span>
<span id="cb12-3"><a href="#cb12-3" tabindex="-1"></a><span class="co"># The indices of EUR individuals</span></span>
<span id="cb12-4"><a href="#cb12-4" tabindex="-1"></a><span class="at">SNPVec=</span>LDL_Locus<span class="sc">$</span>SNP</span>
<span id="cb12-5"><a href="#cb12-5" tabindex="-1"></a><span class="fu">write.table</span>(SNPVec,<span class="fu">glue</span>(<span class="st">"{filepath}/BED_file/LDLSNP.txt"</span>),<span class="at">row.names=</span>F,<span class="at">quote=</span>F,<span class="at">sep=</span><span class="st">"</span><span class="sc">\t</span><span class="st">"</span>,<span class="at">col.names=</span>F)</span>
<span id="cb12-6"><a href="#cb12-6" tabindex="-1"></a><span class="fu">setwd</span>(<span class="st">"~/Plink/"</span>)</span>
<span id="cb12-7"><a href="#cb12-7" tabindex="-1"></a><span class="fu">system</span>(<span class="fu">glue</span>(<span class="st">"./plink --bfile {filepath}/UKB_Genotype_7M/7M_{Chr} --extract {filepath}/BED_file/LDLSNP.txt --make-bed --out {filepath}/BED_file/LDLSNP_Missing"</span>),<span class="at">intern=</span>T)</span>
<span id="cb12-8"><a href="#cb12-8" tabindex="-1"></a><span class="fu">system</span>(<span class="fu">glue</span>(<span class="st">"./plink --bfile {filepath}/BED_file/LDLSNP_Missing --fill-missing-a2 --make-bed --out {filepath}/BED_file/LDLSNP"</span>),<span class="at">intern=</span>T)</span>
<span id="cb12-9"><a href="#cb12-9" tabindex="-1"></a><span class="co"># Imputate very few missing data using referance alleles</span></span>
<span id="cb12-10"><a href="#cb12-10" tabindex="-1"></a><span class="fu">system</span>(<span class="fu">glue</span>(<span class="st">"rm {filepath}/BED_file/LDLSNP_Missing*"</span>))</span>
<span id="cb12-11"><a href="#cb12-11" tabindex="-1"></a><span class="at">geno=</span><span class="fu">BEDMatrix</span>(<span class="fu">glue</span>(<span class="st">"{filepath}/BED_file/LDLSNP.bed"</span>))</span>
<span id="cb12-12"><a href="#cb12-12" tabindex="-1"></a><span class="at">geno_bim=</span><span class="fu">fread</span>(<span class="fu">glue</span>(<span class="st">"{filepath}/BED_file/LDLSNP.bim"</span>))<span class="sc">%&gt;%</span></span>
<span id="cb12-13"><a href="#cb12-13" tabindex="-1"></a>         dplyr<span class="sc">::</span><span class="fu">select</span>(<span class="at">CHR=</span>V1,<span class="at">SNP=</span>V2,<span class="at">BP=</span>V4,<span class="at">A1=</span>V5,<span class="at">A2=</span>V6)<span class="sc">%&gt;%</span><span class="fu">mutate</span>(<span class="at">MarkerName=</span><span class="fu">paste0</span>(SNP,<span class="st">"_"</span>,A1))</span>
<span id="cb12-14"><a href="#cb12-14" tabindex="-1"></a><span class="at">X=</span><span class="fu">as.matrix</span>(geno)</span>
<span id="cb12-15"><a href="#cb12-15" tabindex="-1"></a><span class="at">X=</span>X[EURIND,]</span>
<span id="cb12-16"><a href="#cb12-16" tabindex="-1"></a><span class="at">R=</span><span class="fu">cor</span>(X)</span>
<span id="cb12-17"><a href="#cb12-17" tabindex="-1"></a><span class="fu">colnames</span>(R)<span class="ot">=</span><span class="fu">rownames</span>(R)<span class="ot">=</span>geno_bim<span class="sc">$</span>SNP</span>
<span id="cb12-18"><a href="#cb12-18" tabindex="-1"></a><span class="at">R=</span>R[LDL_Locus<span class="sc">$</span>SNP,LDL_Locus<span class="sc">$</span>SNP]</span>
<span id="cb12-19"><a href="#cb12-19" tabindex="-1"></a><span class="at">R=</span><span class="fu">cov2cor</span>(R)</span></code></pre></div>
</div>
<div class="section level2">
<h2 id="perform-tgvis">Perform TGVIS<a class="anchor" aria-label="anchor" href="#perform-tgvis"></a>
</h2>
<p>Next, we use the function make_design_matrix to convert the eQTL data
into an
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>m</mi><mo>×</mo><mi>p</mi></mrow><annotation encoding="application/x-tex">m\times p</annotation></semantics></math>
matrix, where
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>m</mi><annotation encoding="application/x-tex">m</annotation></semantics></math>
is the number of SNPs and
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>p</mi><annotation encoding="application/x-tex">p</annotation></semantics></math>
is the number of gene-tissue pairs. Then we align the LD matrix, GWAS
data, and eQTL data, and finally perform TGVIS:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bXest</span><span class="op">=</span><span class="fu">TGVIS</span><span class="fu">::</span><span class="fu"><a href="../reference/make_design_matrix.html">make_design_matrix</a></span><span class="op">(</span><span class="va">eQTL</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"SNP"</span>,<span class="st">"Variable"</span>,<span class="st">"Zscore"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">bXest</span><span class="op">=</span><span class="va">bXest</span><span class="op">[</span>,<span class="op">!</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">bXest</span><span class="op">)</span><span class="op">==</span><span class="st">"&lt;NA&gt;"</span><span class="op">)</span><span class="op">]</span></span>
<span><span class="va">bXest</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/NA.html" class="external-link">is.na</a></span><span class="op">(</span><span class="va">bXest</span><span class="op">)</span><span class="op">]</span><span class="op">=</span><span class="fl">0</span></span>
<span><span class="va">bXest</span><span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">bXest</span><span class="op">[</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">SNP</span>,<span class="op">]</span><span class="op">)</span></span>
<span><span class="va">by</span><span class="op">=</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">Zscore</span></span>
<span><span class="va">R</span><span class="op">=</span><span class="va">R</span><span class="op">[</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">SNP</span>,<span class="va">LDL_Locus</span><span class="op">$</span><span class="va">SNP</span><span class="op">]</span></span>
<span><span class="va">t1</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">fit1</span><span class="op">=</span><span class="fu"><a href="../reference/tgvis.html">tgvis</a></span><span class="op">(</span>estimate_inf<span class="op">=</span><span class="cn">F</span>,by<span class="op">=</span><span class="va">by</span>,bXest<span class="op">=</span><span class="va">bXest</span>,LD<span class="op">=</span><span class="va">R</span>,L_vec<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">3</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span>,</span>
<span>          Noutcome<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html" class="external-link">mean</a></span><span class="op">(</span><span class="va">LDL_Locus</span><span class="op">$</span><span class="va">N</span><span class="op">)</span>,var_inf<span class="op">=</span><span class="va">var_inf</span>,</span>
<span>          residual_variance<span class="op">=</span><span class="va">fitldsc</span><span class="op">$</span><span class="va">intercept</span><span class="op">)</span> </span>
<span><span class="va">t2</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Sys.time.html" class="external-link">Sys.time</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/difftime.html" class="external-link">difftime</a></span><span class="op">(</span><span class="va">t2</span>, <span class="va">t1</span>, units <span class="op">=</span> <span class="st">"secs"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="co">#&gt; Time difference of 102.8224 secs</span></span>
<span><span class="co"># residual_variance is the inflaction factor</span></span>
<span><span class="va">fit1</span><span class="op">$</span><span class="va">Summary</span></span>
<span><span class="co">#&gt;               variable cs   cs.logBF    cs.pratt    cs.pip                xqtl</span></span>
<span><span class="co">#&gt; 1           rs11591147  1 509.513399 0.462411150 1.0000000          rs11591147</span></span>
<span><span class="co">#&gt; 2    PCSK9*Whole_Blood  2  87.770503 0.179166944 1.0000000 rs12117661~rs693668</span></span>
<span><span class="co">#&gt; 3            rs2479404  3  25.889077 0.047629044 0.9998935           rs2479404</span></span>
<span><span class="co">#&gt; 4            rs2495489  3  25.889077 0.047629044 0.9998935           rs2495489</span></span>
<span><span class="co">#&gt; 5            rs2495491  3  25.889077 0.047629044 0.9998935           rs2495491</span></span>
<span><span class="co">#&gt; 6          rs150119739  4  12.995401 0.020812107 0.9999856         rs150119739</span></span>
<span><span class="co">#&gt; 7           rs11206517  5  12.730242 0.035418442 0.9529990          rs11206517</span></span>
<span><span class="co">#&gt; 8             rs505151  5  12.730242 0.035418442 0.9529990            rs505151</span></span>
<span><span class="co">#&gt; 9           rs72911441  5  12.730242 0.035418442 0.9529990          rs72911441</span></span>
<span><span class="co">#&gt; 10 TMEM59*Artery_Aorta  6  13.431477 0.033762424 0.9999895          rs12732125</span></span>
<span><span class="co">#&gt; 11     TMEM59*Pancreas  6  13.431477 0.033762424 0.9999895          rs12732125</span></span>
<span><span class="co">#&gt; 12       TMEM59*Testis  6  13.431477 0.033762424 0.9999895          rs12732125</span></span>
<span><span class="co">#&gt; 13          rs12732125  6  13.431477 0.033762424 0.9999895          rs12732125</span></span>
<span><span class="co">#&gt; 14         rs115465289  7   6.489294 0.008102414 0.9837424         rs115465289</span></span>
<span><span class="co">#&gt; 15           rs2479411  8   7.261498 0.002927722 0.9841316           rs2479411</span></span>
<span><span class="co">#&gt; 16           rs2495479  8   7.261498 0.002927722 0.9841316           rs2495479</span></span>
<span><span class="co">#&gt; 17           rs2495480  8   7.261498 0.002927722 0.9841316           rs2495480</span></span>
<span><span class="co">#&gt; 18           rs2495481  8   7.261498 0.002927722 0.9841316           rs2495481</span></span>
<span><span class="co">#&gt; 19           rs2495482  8   7.261498 0.002927722 0.9841316           rs2495482</span></span>
<span><span class="co">#&gt; 20           rs4275490  8   7.261498 0.002927722 0.9841316           rs4275490</span></span>
<span><span class="co">#&gt;    type    estimate        pip        pratt</span></span>
<span><span class="co">#&gt; 1   SNP -71.3590248 1.00000000 0.4624111497</span></span>
<span><span class="co">#&gt; 2  Gene   5.4854726 1.00000000 0.1791669445</span></span>
<span><span class="co">#&gt; 3   SNP  -7.7554503 0.37601913 0.0178352964</span></span>
<span><span class="co">#&gt; 4   SNP  -5.9112270 0.28715670 0.0136748475</span></span>
<span><span class="co">#&gt; 5   SNP  -6.9395350 0.33671763 0.0161189004</span></span>
<span><span class="co">#&gt; 6   SNP  13.3924307 0.99998556 0.0208121069</span></span>
<span><span class="co">#&gt; 7   SNP  16.0798824 0.83744389 0.0312175904</span></span>
<span><span class="co">#&gt; 8   SNP  -1.9671011 0.10371794 0.0037863256</span></span>
<span><span class="co">#&gt; 9   SNP   0.2191409 0.01183717 0.0004145257</span></span>
<span><span class="co">#&gt; 10 Gene   4.2205273 0.24999739 0.0084406060</span></span>
<span><span class="co">#&gt; 11 Gene   2.1536589 0.24999739 0.0084406060</span></span>
<span><span class="co">#&gt; 12 Gene   3.4831624 0.24999739 0.0084406060</span></span>
<span><span class="co">#&gt; 13  SNP  -3.1406787 0.24999739 0.0084406060</span></span>
<span><span class="co">#&gt; 14  SNP -10.3827277 0.98374239 0.0081024141</span></span>
<span><span class="co">#&gt; 15  SNP  -1.8953990 0.15220679 0.0004395556</span></span>
<span><span class="co">#&gt; 16  SNP  -1.6333809 0.13147026 0.0003729371</span></span>
<span><span class="co">#&gt; 17  SNP   1.5374061 0.12385232 0.0003775069</span></span>
<span><span class="co">#&gt; 18  SNP   2.0241300 0.16185945 0.0005098462</span></span>
<span><span class="co">#&gt; 19  SNP   2.1639230 0.17275606 0.0005432653</span></span>
<span><span class="co">#&gt; 20  SNP  -3.0409528 0.24198672 0.0006846112</span></span></code></pre></div>
<p>From the results, we can see that TGVIS first identified a missense
variant in PCSK9, rs11591147, with a Pratt index of 0.462. This SNP is
the PCSK9 variant initially reported by <a href="https://www.nejm.org/doi/full/10.1056/NEJMc0707445" class="external-link">NEJM</a>.
TGVIS also identified the PCSK9-Whole_Blood gene-tissue pair, which has
two causal eQTLs with a Pratt index of 0.179. TGVIS additionally
identified multiple credible sets with Pratt indices not exceeding 0.05.
However, PCSK9 is one of the most significant loci for LDL-C, so even
though the Pratt index is low, this is only relative to this locus.
Cross-loci comparison using Pratt indices is not valid.</p>
</div>
<div class="section level2">
<h2 id="plot-tgivs-results">Plot TGIVS results<a class="anchor" aria-label="anchor" href="#plot-tgivs-results"></a>
</h2>
<p>We also created the plot_tgvis function, which can generate GWAS
locus-zoom plots and TGVIS Pratt index plots:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_tgvis.html">plot_tgvis</a></span><span class="op">(</span>gwas_df<span class="op">=</span><span class="va">LDL_Locus</span>, summary_df<span class="op">=</span><span class="va">fit1</span><span class="op">$</span><span class="va">Summary</span>, </span>
<span>                 cs.pratt_thres <span class="op">=</span> <span class="fl">0.1</span>, y <span class="op">=</span> <span class="st">"z"</span><span class="op">)</span></span>
<span><span class="va">p2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/plot_tgvis.html">plot_tgvis</a></span><span class="op">(</span>gwas_df<span class="op">=</span><span class="va">LDL_Locus</span>, summary_df<span class="op">=</span><span class="va">fit1</span><span class="op">$</span><span class="va">Summary</span>, </span>
<span>                 cs.pratt_thres <span class="op">=</span> <span class="fl">0.1</span>, y <span class="op">=</span> <span class="st">"cs.pratt"</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p1</span></span></code></pre></div>
<p><img src="Analysis_files/figure-html/unnamed-chunk-16-1.png" width="672"></p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">p2</span></span></code></pre></div>
<p><img src="Analysis_files/figure-html/unnamed-chunk-17-1.png" width="672"></p>
<p>In this plot, each circle or square represents a variant. Squares
indicate causal xQTLs, while circles indicate non-causal ones. A single
gene–tissue pair may have multiple causal xQTLs, so multiple signals for
the same pair can appear in the plot. This does not mean there are
multiple genes, but rather multiple xQTLs associated with one
gene–tissue pair. Only variants or gene–tissue pairs with cs.pratt
values exceeding the threshold are annotated.</p>
</div>
<div class="section level2">
<h2 id="caveat-the-use-of-pratt-index">Caveat: the use of Pratt index<a class="anchor" aria-label="anchor" href="#caveat-the-use-of-pratt-index"></a>
</h2>
<p>In our ongoing analysis, we found that the Pratt index can be
affected by many factors, with the largest influence being LD mismatch.
For example, the Pratt index may yield very low values for some GWAS
data that are not largely based on UK Biobank, which is because when
there is LD mismatch, we cannot correctly measure the contribution of
each variable. This issue affects not only TGVIS but all fine-mapping
methods. Therefore, we consider the Pratt index not as a measure of
causality or non-causality, but as a measure of relative importance. We
can regard a Pratt index less than 0.05 as relatively unimportant, such
as in the <code>PCSK9</code> example, where rs2495491 is less important
relative to rs11591147.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yihe Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
