<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Data Preparation • TGVIS</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Data Preparation">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.0/css/all.min.css">
<style>
  body { padding-top: 4.5rem; }
  @media (max-width: 768px) { body { padding-top: 5.5rem; } }
  .col-md-8 { max-width: 90%; }
  .container { max-width: 1400px; }
  .page-main { font-size: 1.05rem; line-height: 1.7; }
</style>
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TGVIS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-user-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">User Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-user-guide">
<li><a class="dropdown-item" href="../articles/Introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/Data-Preparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/Analysis.html">Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Statistical-Principle.html">Statistical Principle</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/harryyiheyang/TGVIS"><span class="fa fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Data Preparation</h1>



      <div class="d-none name"><code>Data Preparation.Rmd</code></div>
    </div>



<p><img src="Inputs.png" width="80%" alt="TGVIS Input" style="display: block; margin: auto;"></p>
<div class="section level2">
<h2 id="input-1-gwas-summary-data-for-tgvis">Input 1: GWAS summary data for TGVIS<a class="anchor" aria-label="anchor" href="#input-1-gwas-summary-data-for-tgvis"></a>
</h2>
<p>This section describes how to prepare GWAS summary statistics and LD
reference panels for TGVIS analysis. TGVIS requires GWAS summary-level
data for a single locus as its in the first input. We strongly recommend
performing quality control using <strong>SBayesRC</strong> prior to
running TGVIS. Although SBayesRC is primarily designed for polygenic
prediction using functional annotations and high-density SNPs (&gt;7
million), its two quality control functions are highly effective for
preprocessing GWAS summary data.</p>
<div class="section level3">
<h3 id="quality-control-with-sbayesrc">Quality Control with SBayesRC<a class="anchor" aria-label="anchor" href="#quality-control-with-sbayesrc"></a>
</h3>
<p>Run the following two functions sequentially:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">SBayesRC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SBayesRC/man/tidy.html" class="external-link">tidy</a></span><span class="op">(</span>mafile <span class="op">=</span> <span class="st">"path/to/your_gwas.ma"</span>,LDdir  <span class="op">=</span> <span class="st">"path/to/ld_folder"</span>,</span>
<span>               output <span class="op">=</span> <span class="st">"output_prefix_tidy.ma"</span>,log2file <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">SBayesRC</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/SBayesRC/man/impute.html" class="external-link">impute</a></span><span class="op">(</span>mafile <span class="op">=</span> <span class="st">"output_prefix_tidy.ma"</span>,LDdir  <span class="op">=</span> <span class="st">"path/to/ld_folder"</span>,</span>
<span>                 output <span class="op">=</span> <span class="st">"output_prefix_imp.ma"</span>,log2file <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="input-format-gwas-summary-statistics">Input Format: GWAS Summary Statistics<a class="anchor" aria-label="anchor" href="#input-format-gwas-summary-statistics"></a>
</h3>
<p>The input <code>mafile</code> must follow the COJO format:</p>
<table class="table">
<thead><tr class="header">
<th>SNP</th>
<th>A1</th>
<th>A2</th>
<th>freq</th>
<th>b</th>
<th>se</th>
<th>p</th>
<th>N</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>rs1001</td>
<td>A</td>
<td>G</td>
<td>0.8493</td>
<td>0.0024</td>
<td>0.0055</td>
<td>0.6653</td>
<td>129850</td>
</tr>
<tr class="even">
<td>rs1002</td>
<td>C</td>
<td>G</td>
<td>0.0306</td>
<td>0.0034</td>
<td>0.0115</td>
<td>0.7659</td>
<td>129799</td>
</tr>
<tr class="odd">
<td>rs1003</td>
<td>A</td>
<td>C</td>
<td>0.5128</td>
<td>0.0045</td>
<td>0.0038</td>
<td>0.2319</td>
<td>129830</td>
</tr>
</tbody>
</table>
</div>
<div class="section level3">
<h3 id="ld-reference-panel">LD Reference Panel<a class="anchor" aria-label="anchor" href="#ld-reference-panel"></a>
</h3>
<p>The <code>LDdir</code> should contain precomputed eigen-decomposition
files for each LD block. We provide links to the SBayesRC LD panels for
multiple ancestries and SNP sets: <a href="https://gctbhub.cloud.edu.au/data/SBayesRC/resources/v2.0/LD/Imputed/ukbEUR_Imputed.zip" class="external-link">EUR</a>,
<a href="https://gctbhub.cloud.edu.au/data/SBayesRC/resources/v2.0/LD/Imputed/ukbEAS_Imputed.zip" class="external-link">EAS</a>,
and <a href="https://gctbhub.cloud.edu.au/data/SBayesRC/resources/v2.0/LD/Imputed/ukbAFR_Imputed.zip" class="external-link">AFR</a></p>
<p>We suggest installing this <a href="https://github.com/zhilizheng/SBayesRC/commit/fac67faf7557b2e542e5cd54fd54b7bd50abb156" class="external-link">version</a>
of SBayesRC, because in our tests, the current version is incompatible
with the BH R package and cannot be installed properly.</p>
</div>
</div>
<div class="section level2">
<h2 id="input-2-xqtl-direct-effect-sizes">Input 2: xQTL direct effect sizes<a class="anchor" aria-label="anchor" href="#input-2-xqtl-direct-effect-sizes"></a>
</h2>
<p>The second input to TGVIS is the weights of xQTL effect size (also
known as direct effects). These are not marginal effects; instead, they
represent the direct causal effect of each xQTL on gene expression or
splicing, estimated while accounting for the LD matrix of nearby tagged
variants.</p>
<p>We <strong>strongly recommend</strong> using precomputed weights from
official sources, which are generally derived from individual-level data
and thus more reliable than fine-mapping from summary statistics: <a href="https://storage.googleapis.com/adult-gtex/bulk-qtl/v10/susie-qtl/GTEx_v10_SuSiE_eQTL.tar" class="external-link">GTEx
v10 SuSiE eQTL weights</a> and <a href="https://predictdb.org/" class="external-link">PredictDB</a>. These weights are typically
estimated using fine-mapping methods such as SuSiE on individual-level
genotype and expression data.</p>
</div>
<div class="section level2">
<h2 id="input-3-ld-matrix">Input 3: LD Matrix<a class="anchor" aria-label="anchor" href="#input-3-ld-matrix"></a>
</h2>
<p>We <strong>strongly recommend</strong> using the LD matrices provided
by SBayesRC as input to TGVIS. We have implemented a function to
construct the required LD matrix:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/build_LD_matrix.html">build_LD_matrix</a></span><span class="op">(</span><span class="va">GWAS_Locus</span>, ldDir <span class="op">=</span> <span class="st">"path/to/ld_folder"</span>, <span class="va">snpinfo</span><span class="op">)</span></span></code></pre></div>
<ul>
<li><p>GWAS_Locus: A data frame containing GWAS summary data for a
single locus (we recommend fewer than 10,000 SNPs to avoid memory
issues).</p></li>
<li><p>ldDir: Path to the directory containing SBayesRC precomputed LD
blocks.</p></li>
<li><p>snpinfo: SNP annotation information, included in the
ancestry-specific LD matrices provided by SBayesRC.</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="input-4-per-snp-heritability-and-inflation-factor">Input 4: Per-SNP Heritability and Inflation Factor<a class="anchor" aria-label="anchor" href="#input-4-per-snp-heritability-and-inflation-factor"></a>
</h2>
<p>The original TGVIS and SuSiE-inf estimate an infinitesimal effect at
the current locus, with its variance estimated via REML. Motivated by
works from Dr. Can Yang’s group, including <a href="https://www.pnas.org/doi/10.1073/pnas.2106858119" class="external-link">MR-APSS</a> and
<a href="https://www.nature.com/articles/s41467-023-42614-7" class="external-link">XMAP</a>,
we note that the variance of the infinitesimal effect can be more
robustly estimated genome-wide using LDSC, rather than locus-specific
estimation. Dr. Yang also pointed out that for most GWAS, the variance
of Z-scores is not 1, but inflated due to uncontrolled confounders. This
inflation factor is precisely the intercept estimated by LDSC.</p>
<p>We have developed an R package <code>ldscR</code>, which includes
ancestry-specific LDSC weights derived from LD matrices prodivded by
PRS-CSX (<a href="https://github.com/getian107/PRScsx" class="external-link">UK Biobank</a>).
Usage is as follows:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://remotes.r-lib.org/reference/install_github.html" class="external-link">install_github</a></span><span class="op">(</span><span class="st">"harryyiheyang/ldscR"</span><span class="op">)</span></span>
<span><span class="fu">ldscR</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ldscR/man/ldsc.univ.html" class="external-link">ldsc.univ</a></span><span class="op">(</span>gwas <span class="op">=</span> <span class="va">GWAS_Locus</span>,LDSC <span class="op">=</span> <span class="va">Hapmap3_EURLDSC</span>,sampling.time <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span><span class="co"># setting sampling.time = 0 disables standard error estimation via blockwise bootstrap.</span></span></code></pre></div>
<p>In addition, empirical observations show that total heritability
typically ranges between 0.1 and 0.3 across traits. Hence, a robust
default is to set: per-SNP heritability = 0.2 / 7.35e6 and inflation factor
= 1.1 (slightly larger than 1).</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yihe Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
