<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>modified_predixcan: Function to estimate causal effects using a modified PrediXcan approach with BIC-based tuning parameter selection — modified_predixcan • TGVIS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="modified_predixcan: Function to estimate causal effects using a modified PrediXcan approach with BIC-based tuning parameter selection — modified_predixcan"><meta name="description" content="This function implements a modified version of the PrediXcan method, using an alternating direction method of multipliers (ADMM) algorithm and BIC for selecting the optimal tuning parameter. It estimates the causal effects of tissue-gene pairs while accounting for direct causal variants."><meta property="og:description" content="This function implements a modified version of the PrediXcan method, using an alternating direction method of multipliers (ADMM) algorithm and BIC for selecting the optimal tuning parameter. It estimates the causal effects of tissue-gene pairs while accounting for direct causal variants."><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.0/css/all.min.css"><style>
  body { padding-top: 4.5rem; }
  @media (max-width: 768px) { body { padding-top: 5.5rem; } }
  .col-md-8 { max-width: 90%; }
  .container { max-width: 1400px; }
  .page-main { font-size: 1.05rem; line-height: 1.7; }
</style></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TGVIS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-user-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">User Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-user-guide"><li><a class="dropdown-item" href="../articles/Introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/Data-Preparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/Analysis.html">Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Statistical-Principle.html">Statistical Principle</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/harryyiheyang/TGVIS"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>modified_predixcan: Function to estimate causal effects using a modified PrediXcan approach with BIC-based tuning parameter selection</h1>

      <div class="d-none name"><code>modified_predixcan.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function implements a modified version of the PrediXcan method, using an alternating direction method of multipliers (ADMM) algorithm and BIC for selecting the optimal tuning parameter. It estimates the causal effects of tissue-gene pairs while accounting for direct causal variants.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">modified_predixcan</span><span class="op">(</span></span>
<span>  <span class="va">by</span>,</span>
<span>  <span class="va">bxest</span>,</span>
<span>  <span class="va">LD</span>,</span>
<span>  tauvec <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq</a></span><span class="op">(</span><span class="fl">3</span>, <span class="fl">10</span>, by <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>,</span>
<span>  rho.gamma <span class="op">=</span> <span class="fl">1.5</span>,</span>
<span>  max.iter <span class="op">=</span> <span class="fl">15</span>,</span>
<span>  max.eps <span class="op">=</span> <span class="fl">0.005</span>,</span>
<span>  ebic.factor <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  normmax <span class="op">=</span> <span class="fl">2</span>,</span>
<span>  pleiotropy.rm <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-by">by<a class="anchor" aria-label="anchor" href="#arg-by"></a></dt>
<dd><p>A vector of Z-scores of marginal effects from the outcome GWAS (same as in ctwas).</p></dd>


<dt id="arg-bxest">bxest<a class="anchor" aria-label="anchor" href="#arg-bxest"></a></dt>
<dd><p>A vector of direct xQTL effect estimates for a tissue-gene pair.</p></dd>


<dt id="arg-ld">LD<a class="anchor" aria-label="anchor" href="#arg-ld"></a></dt>
<dd><p>The LD matrix of variants (same as in ctwas).</p></dd>


<dt id="arg-tauvec">tauvec<a class="anchor" aria-label="anchor" href="#arg-tauvec"></a></dt>
<dd><p>A vector of tuning parameters used in penalizing the direct causal effect. Default is `seq(3,10,by=1)`.</p></dd>


<dt id="arg-rho-gamma">rho.gamma<a class="anchor" aria-label="anchor" href="#arg-rho-gamma"></a></dt>
<dd><p>A parameter set in the ADMM algorithm. Default is 1.5.</p></dd>


<dt id="arg-max-iter">max.iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>The maximum number of iterations for the ADMM algorithm. Default is 15.</p></dd>


<dt id="arg-max-eps">max.eps<a class="anchor" aria-label="anchor" href="#arg-max-eps"></a></dt>
<dd><p>The convergence tolerance for the ADMM algorithm. Default is 0.005.</p></dd>


<dt id="arg-ebic-factor">ebic.factor<a class="anchor" aria-label="anchor" href="#arg-ebic-factor"></a></dt>
<dd><p>The extended BIC factor for model selection. Default is 2.</p></dd>


<dt id="arg-pleiotropy-rm">pleiotropy.rm<a class="anchor" aria-label="anchor" href="#arg-pleiotropy-rm"></a></dt>
<dd><p>A vector of indices specifying which variants should not be considered as having direct causal effects.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p>
<dl><dt>theta</dt>
<dd><p>The estimated effect size of the tissue-gene pair.</p></dd>

<dt>gamma</dt>
<dd><p>The estimated effect sizes of the direct causal variants.</p></dd>

<dt>covtheta</dt>
<dd><p>The covariance of the estimated effect size `theta`.</p></dd>

<dt>Bic</dt>
<dd><p>The BIC values for each tuning parameter.</p></dd>

<dt>Btheta</dt>
<dd><p>The estimated `theta` values for each tuning parameter.</p></dd>

<dt>Bgamma</dt>
<dd><p>The estimated `gamma` values for each tuning parameter.</p></dd>

<dt>Eta</dt>
<dd><p>The estimated linear predictor for the tissue-gene pair.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yihe Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

