<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><title>tgvis: Function to estimate and select the optimal number of single effects using profile-likelihood and BIC — tgvis • TGVIS</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet"><link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet"><script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="tgvis: Function to estimate and select the optimal number of single effects using profile-likelihood and BIC — tgvis"><meta name="description" content="This function estimates the number of single effects in a locus by combining profile-likelihood methods and Bayesian Information Criterion (BIC) to optimize the model. It includes resampling for estimating standard errors and performs score tests for infinitesimal effects."><meta property="og:description" content="This function estimates the number of single effects in a locus by combining profile-likelihood methods and Bayesian Information Criterion (BIC) to optimize the model. It includes resampling for estimating standard errors and performs score tests for infinitesimal effects."><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.5.0/css/all.min.css"><style>
  body { padding-top: 4.5rem; }
  @media (max-width: 768px) { body { padding-top: 5.5rem; } }
  .col-md-8 { max-width: 90%; }
  .container { max-width: 1400px; }
  .page-main { font-size: 1.05rem; line-height: 1.7; }
</style></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-primary" data-bs-theme="dark" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">TGVIS</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item"><a class="nav-link" href="../index.html">Home</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-user-guide" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">User Guide</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-user-guide"><li><a class="dropdown-item" href="../articles/Introduction.html">Introduction</a></li>
    <li><a class="dropdown-item" href="../articles/Data-Preparation.html">Data Preparation</a></li>
    <li><a class="dropdown-item" href="../articles/Analysis.html">Analysis</a></li>
    <li><a class="dropdown-item" href="../articles/Statistical-Principle.html">Statistical Principle</a></li>
  </ul></li>
<li class="active nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul><ul class="navbar-nav"><li class="nav-item"><a class="external-link nav-link" href="https://github.com/harryyiheyang/TGVIS"><span class="fa fa-github fa-lg"></span></a></li>
      </ul></div>


  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>tgvis: Function to estimate and select the optimal number of single effects using profile-likelihood and BIC</h1>

      <div class="d-none name"><code>tgvis.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>This function estimates the number of single effects in a locus by combining profile-likelihood methods and Bayesian Information Criterion (BIC) to optimize the model. It includes resampling for estimating standard errors and performs score tests for infinitesimal effects.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tgvis</span><span class="op">(</span></span>
<span>  estimate_inf <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  <span class="va">by</span>,</span>
<span>  <span class="va">bXest</span>,</span>
<span>  <span class="va">LD</span>,</span>
<span>  <span class="va">Noutcome</span>,</span>
<span>  L_vec <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">)</span>,</span>
<span>  var_inf <span class="op">=</span> <span class="fl">1e-07</span>,</span>
<span>  estimate_residual_variance <span class="op">=</span> <span class="cn">F</span>,</span>
<span>  scaled_prior_variance <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  residual_variance <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  max_iter <span class="op">=</span> <span class="fl">50</span>,</span>
<span>  max_eps <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  susie_iter <span class="op">=</span> <span class="fl">500</span>,</span>
<span>  pip_thres_cred <span class="op">=</span> <span class="fl">0.95</span>,</span>
<span>  eigen_thres <span class="op">=</span> <span class="fl">0.999</span>,</span>
<span>  varinf_upper_boundary <span class="op">=</span> <span class="fl">0.25</span>,</span>
<span>  varinf_lower_boundary <span class="op">=</span> <span class="fl">0.001</span>,</span>
<span>  ebic_beta <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  ebic_upsilon <span class="op">=</span> <span class="fl">1</span>,</span>
<span>  pip_min <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pv_thres <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  pleiotropy_rm <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior_weight_theta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  prior_weight_gamma <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  standization <span class="op">=</span> <span class="cn">T</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>


<dl><dt id="arg-estimate-inf">estimate_inf<a class="anchor" aria-label="anchor" href="#arg-estimate-inf"></a></dt>
<dd><p>An indicator of whether estimating the infinitesimal effect. Default is F.</p></dd>


<dt id="arg-by">by<a class="anchor" aria-label="anchor" href="#arg-by"></a></dt>
<dd><p>A vector of Z-scores of the marginal effects from the outcome GWAS.</p></dd>


<dt id="arg-bxest">bXest<a class="anchor" aria-label="anchor" href="#arg-bxest"></a></dt>
<dd><p>A matrix of direct effect estimates based on the Z-scores of tissue-gene pairs.</p></dd>


<dt id="arg-ld">LD<a class="anchor" aria-label="anchor" href="#arg-ld"></a></dt>
<dd><p>The LD matrix of variants.</p></dd>


<dt id="arg-noutcome">Noutcome<a class="anchor" aria-label="anchor" href="#arg-noutcome"></a></dt>
<dd><p>The sample size of the outcome GWAS.</p></dd>


<dt id="arg-l-vec">L_vec<a class="anchor" aria-label="anchor" href="#arg-l-vec"></a></dt>
<dd><p>A vector of candidate numbers of single effects used in BIC. Default is `c(1:8)`.</p></dd>


<dt id="arg-var-inf">var_inf<a class="anchor" aria-label="anchor" href="#arg-var-inf"></a></dt>
<dd><p>When estimate_inf = F, the variance of infinitesimal effect (estimated by LDSC possibly). Default is 1e-7.</p></dd>


<dt id="arg-estimate-residual-variance">estimate_residual_variance<a class="anchor" aria-label="anchor" href="#arg-estimate-residual-variance"></a></dt>
<dd><p>An indicator of whether of not estimating the variance of residuals in SuSiE. Default is F.</p></dd>


<dt id="arg-scaled-prior-variance">scaled_prior_variance<a class="anchor" aria-label="anchor" href="#arg-scaled-prior-variance"></a></dt>
<dd><p>The prior variance of signals in SuSiE. Default is 0.5 which is slightly larger than 0.2 in SuSiE software.</p></dd>


<dt id="arg-residual-variance">residual_variance<a class="anchor" aria-label="anchor" href="#arg-residual-variance"></a></dt>
<dd><p>The residual variance.  Default is 1.</p></dd>


<dt id="arg-max-iter">max_iter<a class="anchor" aria-label="anchor" href="#arg-max-iter"></a></dt>
<dd><p>The maximum number of iterations for the profile-likelihood algorithm. Default is 50.</p></dd>


<dt id="arg-max-eps">max_eps<a class="anchor" aria-label="anchor" href="#arg-max-eps"></a></dt>
<dd><p>The convergence tolerance for the profile-likelihood algorithm. Default is 1e-3.</p></dd>


<dt id="arg-susie-iter">susie_iter<a class="anchor" aria-label="anchor" href="#arg-susie-iter"></a></dt>
<dd><p>The maximum number of iterations for `susie_rss` within the profile-likelihood algorithm. Default is 50.</p></dd>


<dt id="arg-pip-thres-cred">pip_thres_cred<a class="anchor" aria-label="anchor" href="#arg-pip-thres-cred"></a></dt>
<dd><p>The cumulative PIP threshold for variables in a credible set. Default is 0.95.</p></dd>


<dt id="arg-eigen-thres">eigen_thres<a class="anchor" aria-label="anchor" href="#arg-eigen-thres"></a></dt>
<dd><p>The threshold of eigenvalues for modelling the infinitesimal effect. Default is 1.</p></dd>


<dt id="arg-varinf-upper-boundary">varinf_upper_boundary<a class="anchor" aria-label="anchor" href="#arg-varinf-upper-boundary"></a></dt>
<dd><p>The upper boundary for the prior variance of infinitesimal effects, multiplied by var(y) to adapt to different locus variances. Default is 0.25.</p></dd>


<dt id="arg-varinf-lower-boundary">varinf_lower_boundary<a class="anchor" aria-label="anchor" href="#arg-varinf-lower-boundary"></a></dt>
<dd><p>The lower boundary for the prior variance of infinitesimal effects, not multiplied by var(y). Default is 0.001.</p></dd>


<dt id="arg-ebic-beta">ebic_beta<a class="anchor" aria-label="anchor" href="#arg-ebic-beta"></a></dt>
<dd><p>The extended BIC factor for causal effects of tissue-gene pairs and direct causal variants used in BIC computation. Default is 1.</p></dd>


<dt id="arg-ebic-upsilon">ebic_upsilon<a class="anchor" aria-label="anchor" href="#arg-ebic-upsilon"></a></dt>
<dd><p>The extended BIC factor for infinitesimal effects used in BIC computation. Default is 1.</p></dd>


<dt id="arg-pip-min">pip_min<a class="anchor" aria-label="anchor" href="#arg-pip-min"></a></dt>
<dd><p>The minimum PIP threshold for individual causal effects in the profile-likelihood. This is used to specify which tissue-gene pairs and direct causal variants to include in the score test of variance of infinitesimal effects. Default is 0.05.</p></dd>


<dt id="arg-pv-thres">pv_thres<a class="anchor" aria-label="anchor" href="#arg-pv-thres"></a></dt>
<dd><p>The p-value threshold for the score test. Default is 0.05.</p></dd>


<dt id="arg-pleiotropy-rm">pleiotropy_rm<a class="anchor" aria-label="anchor" href="#arg-pleiotropy-rm"></a></dt>
<dd><p>A vector of indices specifying which variants should not be considered as having direct causal effects.</p></dd>


<dt id="arg-prior-weight-theta">prior_weight_theta<a class="anchor" aria-label="anchor" href="#arg-prior-weight-theta"></a></dt>
<dd><p>A vector of prior weights of gene-tissue pairs, which will be used as input in SuSiE. Default is <code>NULL</code>.</p></dd>


<dt id="arg-prior-weight-gamma">prior_weight_gamma<a class="anchor" aria-label="anchor" href="#arg-prior-weight-gamma"></a></dt>
<dd><p>A vector of prior weights of direct causal variants, which will be used as input in SuSiE. Default is <code>NULL</code>.</p></dd>


<dt id="arg-standization">standization<a class="anchor" aria-label="anchor" href="#arg-standization"></a></dt>
<dd><p>A indicator of whether standardizing the input when performing SuSiE for fine-mapping causal gene-tissue pairs and direct causal variants. Default is <code>T</code>.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    <p>A list containing:</p>
<dl><dt>theta</dt>
<dd><p>The estimated effects for tissue-gene pairs, scaled by the outcome GWAS sample size.</p></dd>

<dt>gamma</dt>
<dd><p>The estimated effects for direct causal variants, scaled by the outcome GWAS sample size.</p></dd>

<dt>theta.pip</dt>
<dd><p>Posterior inclusion probabilities (PIP) for tissue-gene pairs.</p></dd>

<dt>gamma.pip</dt>
<dd><p>Posterior inclusion probabilities (PIP) for direct causal variants.</p></dd>

<dt>theta.pratt</dt>
<dd><p>Pratt estimations for tissue-gene pairs.</p></dd>

<dt>gamma.pratt</dt>
<dd><p>Pratt estimations for direct causal variants.</p></dd>

<dt>theta.cs</dt>
<dd><p>Credible set indicators for tissue-gene pairs.</p></dd>

<dt>gamma.cs</dt>
<dd><p>Credible set indicators for direct causal variants.</p></dd>

<dt>theta.cs.pip</dt>
<dd><p>PIP within credible sets for tissue-gene pairs.</p></dd>

<dt>gamma.cs.pip</dt>
<dd><p>PIP within credible sets for direct causal variants.</p></dd>

<dt>upsilon</dt>
<dd><p>The estimated infinitesimal effects.</p></dd>

<dt>var.upsilon</dt>
<dd><p>The estimated variance of infinitesimal effects.</p></dd>

<dt>fit.causal</dt>
<dd><p>The SuSiE fit object for the causal analysis.</p></dd>

<dt>cs.summary</dt>
<dd><p>A summary of the credible sets obtained from the analysis.</p></dd>

<dt>Bicvec</dt>
<dd><p>A vector of BIC values for each candidate number of single effects.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Yihe Yang.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer></div>





  </body></html>

